java-pkg-simple.eclass: Allow compilation of multiple modules in one pkg

2Kmm opened the pull request:
Auto generate parameter `--module-source-path` if needed.

Closes: https://bugs.gentoo.org/925691

2Kmm commented:
I am not exactly sure, how the policy is. Should I revbump the fixed ebuilds?

vaukai commented:
> I am not exactly sure, how the policy is. Should I revbump the fixed ebuilds?

Sorry, it looks like you've worked on the old approach still producing the "ignoreme.jar" file.  Would it also work for the newer approach using `JAVADOC_SRC_DIRS` like the upcoming [byte-buddy](https://github.com/gentoo/gentoo/pull/33579/commits/bf2490ee707e66d57ee745175dbfe9dae2d20c36#diff-50d3f146f0461b0d0ff06d705de9d849c34b2eee0d64e7726a475ff5c5258738)?

2Kmm commented:
> Sorry, it looks like you've worked on the old approach still producing the "ignoreme.jar" file. Would it also work for the newer approach using `JAVADOC_SRC_DIRS` like the upcoming [byte-buddy](https://github.com/gentoo/gentoo/pull/33579/commits/bf2490ee707e66d57ee745175dbfe9dae2d20c36#diff-50d3f146f0461b0d0ff06d705de9d849c34b2eee0d64e7726a475ff5c5258738)?

Yes, my work is mostly decoupled from your work. Mainly, it obsoletes the need for a `JAVA_MODULE_SOURCE_PATH` variable because it auto-generates it. So our changes should work nicely together.


> Back to your question. Think of a user having `cdi-api` or `jaxb-runtime` already installed. Your changes on the ebuilds are strongly related to that change on the eclass. For the user, the eclass change will run in immediately on next sync. But there is nothing triggering an update of the packages unless their commits contain a revbump.

That is the point why I am asking. For users with USE=-doc, nothing changes, and therefore there is no need to recompile. For users with USE=doc, the ebuilds failed before and therefore should not be installed in the first place. So, a revbump would only trigger an update for those users, that don't need it...

2Kmm commented:
> Only today I took the time to run some tests. It fails with both [cdi-api](https://github.com/gentoo/gentoo/pull/35305) and [byte-buddy](https://github.com/gentoo/gentoo/pull/33579) unless I keep removing `module-info.java`.

Oh, Sorry, this was a missunderstanding. I have only changed the parts in `java-pkg-simple.eclass`, not the corresponding ones in `java-utils-2.eclass` (`ejavadoc`), as I understand that you wanted to move this into `java-pkg-simple.eclass`. So, yes, this does not work as is with the new ebuilds. It demonstrates the approach, and the approach goes hand in hand with your planed changes.

2Kmm commented:
Next try. I think this is like you imagined it, right?

2Kmm commented:
Rebased on current master and updated description of JAVADOC_SRC_DIRS

vaukai commented:
P.S.: 
For removing `JAVADOC_SRC_DIRS` and `JAVADOC_CLASSPATH` from `java-utils-2.eclass` my older commits there can be and should be reverted (git revert ...).

2Kmm commented:
> Another point is the name of the new function `java-pkg-simple_generate_documentation()`. I'd prefer a more apidoc-specific name, maybe `java-pkg-simple_call_ejavadoc`

Renamed :heavy_check_mark: 

> And then, all packages listed in `qgrep JAVADOC_SRC_DIRS` should be switched to the new function name, all in one commit.

I didn't realized, that there where already users of this in the tree. Thanks for hinting. Done :heavy_check_mark: 

> And a question: Is `--module-source-path` per module? From reading [this description](https://www.logicbig.com/tutorials/core-java-tutorial/modules/modes.html) I got the impression we could have it the same in the whole `ebuild`.

It can be both. You can give it a single path, but then there are restrictions. A module has to be in a direct subdir of the module-source-path and the subdir has to be exactly the name of the module. With the other style, you can call it once for every module and give it the exact, arbitrary path to this module. With the later variant, we can auto-generate it (as demonstrated) so the ebuild author doesn't has to bother at all.

> P.S.: For removing `JAVADOC_SRC_DIRS` and `JAVADOC_CLASSPATH` from `java-utils-2.eclass` my older commits there can be and should be reverted (git revert ...).

Personally, I prefer to make commits that allow to follow the flow of the code. So one move, rather then one delete and one add. But if it is your wish (or a gentoo rule?) I can separate this out into different commits. Should I?

2Kmm commented:
I added some more automation. With the last commit, the `src_install` from packages like `cdi-api` or `jaxb-runtime` could be dropped, as `java-pkg-simple_src_install` can handle multiple jars now.

As this is quite a bit of change now and I don't know if this is the way to go, I converted this PR to a draft for now.

2Kmm commented on eclass/java-pkg-simple.eclass:
```diff
@@ -453,18 +565,42 @@ java-pkg-simple_src_compile() {
+
+	# javadoc
+	if has doc ${JAVA_PKG_IUSE} && use doc; then
+		if has binary ${JAVA_PKG_IUSE} && use binary; then
+			ewarn "Generation of documentation is not supported when installing binary."
```
Before, `src_compile` would not generate the documentation, if the `binary` use flag was active. Should we keep this, or should we still use the provided sources to generate it? How about installing the sources, when use flag `sources` is active? I think it should be handled the same as the doc generation. Either we drop both if `binary` is active, or we do both regardless.

2Kmm commented on eclass/java-pkg-simple.eclass:
```diff
@@ -335,19 +412,70 @@ java-pkg-simple_prepend_resources() {
+# @USAGE: java-pkg-simple_call_ejavadoc
 # @DESCRIPTION:
-# src_compile for simple bare source java packages. Finds all *.java
-# sources in ${JAVA_SRC_DIR}, compiles them with the classpath
-# calculated from ${JAVA_GENTOO_CLASSPATH}, and packages the resulting
-# classes to a single ${JAVA_JAR_FILENAME}. If the file
+# Generates documentation
```
Any idea for a better description?

Flowdalic commented on eclass/java-pkg-simple.eclass:
```diff
@@ -335,19 +412,70 @@ java-pkg-simple_prepend_resources() {
+	# if target < 9, we need to compile module-info.java separately
+	# as this feature is not supported before Java 9
+	local target="$(java-pkg_get-target)"
+	if [[ ${target#1.} -lt 9 ]]; then
```
Could probably be simplified to something like

```bash
[[ ${target#1.} -lt 9 ]] && find_args+=(! -name module-info.java)
```

Also, missing `die` on find invocations.

2Kmm replied:
Good idea. Same for the `ejavac` calls. But maybe that should be done in a separate PR, as this was not introduced here and this PR is already quite big. What do you think?

vaukai replied:
> Good idea. Same for the `ejavac`

usually we do not `|| die` after ejavac, ejunit, einfo, emake ...

2Kmm replied:
I meant the first part, the simplification. That can be used for `ejavac`, too. Here:
```bash
    if [[ -z ${moduleinfo} ]] || [[ ${target#1.} -lt 9 ]]; then
        ejavac -d ${classes} -encoding ${JAVA_ENCODING}\
            ${classpath:+-classpath ${classpath}} ${JAVAC_ARGS} @${sources}
    else
        ejavac -d ${classes} "${modulesourcepath[@]}" -encoding ${JAVA_ENCODING}\
            ${classpath:+--module-path ${classpath}} --module-version ${PV}\
            ${JAVAC_ARGS} @${sources}
    fi
```
Use an `ejavac_args` array and only a single call to `ejavac` after the if.

Flowdalic commented on eclass/java-pkg-simple.eclass:
```diff
@@ -335,19 +412,70 @@ java-pkg-simple_prepend_resources() {
+	java-pkg-simple_getmodulesourcepath "${moduleinfo[@]}"
+
+	# create the target directory
+	mkdir -p ${apidoc}
```
```suggestion
	mkdir -p ${apidoc} || die
```

Flowdalic commented on eclass/java-pkg-simple.eclass:
```diff
@@ -373,7 +502,10 @@ java-pkg-simple_src_compile() {
-	moduleinfo=$(find "${JAVA_SRC_DIR[@]}" -name module-info.java)
+	JAVA_INTERNAL_SRC_DIRS+=( "${JAVA_SRC_DIR[@]/#/${PWD}/}" )
+	local -a moduleinfo
+	readarray -t moduleinfo <<<"$(find "${JAVA_SRC_DIR[@]}" -name module-info.java)"
```
```suggestion
	readarray -t moduleinfo <<<"$(find "${JAVA_SRC_DIR[@]}" -name module-info.java || die)"
```

Flowdalic commented on eclass/java-pkg-simple.eclass:
```diff
@@ -453,18 +565,42 @@ java-pkg-simple_src_compile() {
+	# javadoc
+	if has doc ${JAVA_PKG_IUSE} && use doc; then
+		if has binary ${JAVA_PKG_IUSE} && use binary; then
+			ewarn "Generation of documentation is not supported when installing binary."
```
Be clear what "installing binary" means.
```suggestion
			ewarn "Generation of documentation is not supported when installing binary (USE=binary)."
```

Flowdalic commented on eclass/java-pkg-simple.eclass:
```diff
@@ -453,18 +565,42 @@ java-pkg-simple_src_compile() {
+		if has binary ${JAVA_PKG_IUSE} && use binary; then
+			ewarn "Generation of documentation is not supported when installing binary."
+		elif [[ ${JAVADOC_SRC_DIRS} ]]; then
+			einfo "JAVADOC_SRC_DIRS exists, you need to call java-pkg-simple_call_ejavadoc separately"
```
Haven't looked at the logic, but this appears to be a comment towards the ebuild developer and not the Gentoo users. In this case, this should maybe a `die` instead of `einfo`?  Or at least a `eqawarn`.

2Kmm replied:
Before my changes, this was just an information for the developer, that was **always** shown when `JAVADOC_SRC_DIRS` was used. A `die` would have prevented a use of that feature all together. But after my changes one could prevent this notice by calling `java-pkg-simple_compile_jar` instead of `java-pkg-simple_src_compile`. So maybe we should change this indeed to an `eqawarn` and suggest to not use `java-pkg-simple_src_compile` together with `JAVADOC_SRC_DIRS`.

(A `die` would break existing, functioning ebuilds. I would not do this at this point. Maybe after some time with `eqawarn`)

vaukai replied:
> Before my changes, this was just an information for the developer, that was **always** shown when `JAVADOC_SRC_DIRS` was used. A `die` would have prevented a use of that feature all together. But after my changes one could prevent this notice by calling `java-pkg-simple_compile_jar` instead of `java-pkg-simple_src_compile`. So maybe we should change this indeed to an `eqawarn` and suggest to not use `java-pkg-simple_src_compile` together with `JAVADOC_SRC_DIRS`.
> 
> (A `die` would break existing, functioning ebuilds. I would not do this at this point. Maybe after some time with `eqawarn`)

if @fordfrog agrees, [that message](https://github.com/gentoo/gentoo/pull/34108#pullrequestreview-1798750973) could be dropped i guess.

Flowdalic commented on eclass/java-pkg-simple.eclass:
```diff
@@ -542,15 +679,17 @@ java-pkg-simple_src_test() {
 	fi
-	moduleinfo=$(find "${JAVA_TEST_SRC_DIR[@]}" -name module-info.java)
+	local -a moduleinfo
+	readarray -t moduleinfo <<<"$(find "${JAVA_TEST_SRC_DIR[@]}" -name module-info.java)"
```
```suggestion
	readarray -t moduleinfo <<<"$(find "${JAVA_TEST_SRC_DIR[@]}" -name module-info.java || die)"
```

Flowdalic reviewed (commented):
First single-pass review. Haven't looked at the actual logic. Maybe we can use the  opportunity to modernize the API of the eclass regarding the JAVADOC_CLASSPATH variable to use a bash array. However, I am not sure if this is feasible.

2Kmm commented on eclass/java-pkg-simple.eclass:
```diff
@@ -216,6 +216,59 @@ fi
 # @DESCRIPTION:
 # It is almost equivalent to ${JAVA_RESOURCE_DIRS} in src_test.
 
+# @ECLASS_VARIABLE: JAVADOC_CLASSPATH
```
It is a relative new variable with just a few users (qgrep found 5 ebuilds currently). If we want to make this an array, then better now.

vaukai replied:
@Flowdalic 
i'd prefer having it the same way as `JAVA_GENTOO_CLASSPATH`. Why should we have `Comma or space separated list of ...` for the one and `array` for the other? ebuild writers like me would only get confused.

btw: the eclass variables `JAVADOC_CLASSPATH` and `JAVADOC_SRC_DIRS` both already exist and are now getting moved to `java-pkg-simple.eclass`.

2Kmm commented on eclass/java-pkg-simple.eclass:
```diff
@@ -453,18 +565,42 @@ java-pkg-simple_src_compile() {
 			|| die "updating MANIFEST.MF failed"
 		rm -f "${T}/add-to-MANIFEST.MF" || die "cannot remove"
 	fi
+	JAVA_INTERNAL_JAR_FILES+=( "${PWD}/${JAVA_JAR_FILENAME}" )
```
As I learned, there is a use case where a jar-file is generated but should not be installed. So we need a way to tell `java-pkg-simple_src_compile` to not automatically add to `JAVA_INTERNAL_JAR_FILES`.

2Kmm replied:
No, no general support. Just that it stores which jars where compiled and installs all of them in `java-pkg-simple_src_install`. But I can save this part for a later PR.

2Kmm replied:
I have dropped the changes for auto-installing jars. `java-pkg-simple_src_install` is still touched to install the accumulated sources from the `java-pkg-simple_compile_jar` runs. Are you fine with this?

2Kmm replied:
I reduced the changes further to keep it simple. `java-pkg-simple_src_install` is not touched anymore.

vaukai commented on eclass/java-pkg-simple.eclass:
```diff
@@ -455,6 +556,29 @@ java-pkg-simple_src_compile() {
+		if has binary ${JAVA_PKG_IUSE} && use binary; then
+			ewarn "Generation of documentation is not supported when installing binary (USE=binary)."
+		elif [[ ${JAVADOC_SRC_DIRS} ]]; then
+			eqawarn "JAVADOC_SRC_DIRS is set and java-pkg-simple_src_compile called."
+			eqawarn "Please use JAVADOC_SRC_DIRS together with java-pkg-simple_compile_jar/java-pkg-simple_call_ejavadoc instead."
+			eqawarn "This will become an error in the future."
```
> I replaced `einfo` with `eqawarn` and adjusted the message. @fordfrog, @vaukai: Are you fine with the wording?

personally, i'd prefer the long version:
```
"When setting JAVADOC_SRC_DIRS use java-pkg-simple_compile_jar 
instead of java-pkg-simple_src_compile for the jar files 
and call java-pkg-simple_call_ejavadoc for javadoc."
```

2Kmm replied:
Yes, your version reads a lot better. Thanks!

vaukai commented:
@2Kmm please be aware of `dev-java/asm: add 9.7` in https://github.com/gentoo/gentoo/pull/33579/commits.
